src/Data/HashMap/Linear.hs

```diff
@@ -120,141 +149,255 @@ data ProbeResult k v where
-  where
-    extractPairs :: Keyed k =>
-      Int -> RobinArr k v #-> (RobinArr k v, Ur [(k,v)])
-    extractPairs size arr = walk arr (size - 1) []
-
-    walk :: Keyed k =>
-      RobinArr k v #-> Int -> [(k,v)] -> (RobinArr k v, Ur [(k,v)])
-    walk arr ix kvs
-      | ix < 0 = (arr, Ur kvs)
-      | otherwise = Array.read arr ix & \case
-        (arr', Ur Nothing) ->
-          walk arr' (ix-1) kvs
-        (arr', Ur (Just (RobinVal _ k v))) ->
-          walk arr' (ix-1) ((k,v):kvs)
-
--- | 'delete h k' deletes key k and its value if it is present.
--- If it's not present, this does nothing.
+-- | Run a computation with an 'HashMap' containing given key-value pairs.
+fromList :: forall k v b.
+  Keyed k => [(k, v)] -> (HashMap k v #-> Ur b) #-> Ur b
+fromList xs scope =
+  Array.alloc
+    (max
+      1
+      (ceiling @Float @Int (fromIntegral (length xs) / constMaxLoadFactor)))
+    Nothing
+    (\arr -> scope (insertAll xs (HashMap 0 arr)))
+
+-- | The most general modification function; which can insert, update or delete
+-- a value of the key.
+alter :: Keyed k => HashMap k v #-> k -> (Maybe v -> Maybe v) -> HashMap k v
+alter hm key f =
+  idealIndexForKey key hm & \(hm', Ur idx) ->
+    probeFrom (key, 0) idx hm' & \case
+      -- The key does not exist, and there is an empty cell to insert.
+      (HashMap count arr, IndexToInsert psl ix) ->
+        Ur (f Nothing) & \case
+          -- We don't need to insert anything.
+          Ur Nothing -> HashMap count arr
+          -- We need to insert a new key.
+          Ur (Just v)->
+            HashMap
+             (count+1)
+             (Array.write arr ix (Just (RobinVal psl key v)))
+             & growMapIfNecessary
+      -- The key exists.
+      (hm'', IndexToUpdate v psl ix) ->
+        capacity hm'' & \(HashMap count arr, Ur cap) ->
+          Ur (f (Just v)) & \case
+            -- We need to delete it.
+            Ur Nothing ->
+              Array.write arr ix Nothing & \arr' ->
+                shiftSegmentBackward cap arr' ((ix + 1) `mod` cap) & \arr'' ->
+                  HashMap
+                    (count - 1)
+                    arr''
+            -- We need to update it.
+            Ur (Just new)->
+              HashMap
+                count
+                (Array.write arr ix (Just (RobinVal psl key new)))
+      -- The key does not exist, but there is a key to evict.
+      (hm, IndexToSwap evicted psl ix) ->
+        Ur (f Nothing) & \case
+          -- We don't need to insert anything.
+          Ur Nothing -> hm
+          -- We need to insert a new key.
+          Ur (Just v)->
+            capacity hm & \(HashMap count arr, Ur cap) ->
+              tryInsertAtIndex
+                (HashMap
+                  count
+                  (Array.write arr ix (Just (RobinVal psl key v))))
+                ((ix + 1) `mod` cap)
+                (incRobinValPSL evicted)
+              & growMapIfNecessary
```

Comment:
I (politely) think this is a terrible idea. It's much harder to maintain code like this. Implement `alter` in terms of `insert` and `delete` and not the other way around. I'm blocking this PR for this.

Reply:
After some discussion, I'm reconsidering. We can't implement in terms of `insert` and `delete` without re-probing and I can't find a way to re-write `alter` to be as shorter but as efficient as this version.